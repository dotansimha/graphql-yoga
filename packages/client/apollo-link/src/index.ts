import { ExecutorLink } from '@graphql-tools/executor-apollo-link';
import { buildHTTPExecutor, HTTPExecutorOptions } from '@graphql-tools/executor-http';

export type YogaLinkOptions = HTTPExecutorOptions;

export class YogaLink extends ExecutorLink {
  constructor(options: YogaLinkOptions) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    super(buildHTTPExecutor(options) as any);
  }
}
