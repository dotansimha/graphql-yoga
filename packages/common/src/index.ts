export * from './types'
export * from './logger'
export * from './server'

export * from './graphiql'
export * from '@envelop/core'
export * from '@graphql-yoga/subscription'
export * from './plugins/types'
export * from './plugins/useRequestParser'
export { Plugin } from './plugins/types'
