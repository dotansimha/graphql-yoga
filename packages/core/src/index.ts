import { GraphiQLOptions, handleRequest } from '@graphql-yoga/handler'
import { GraphQLSchema } from 'graphql'
import {
  Plugin,
  GetEnvelopedFn,
  envelop,
  useMaskedErrors,
  UseMaskedErrorsOpts,
  useExtendContext,
  enableIf,
  useLogger,
  useSchema,
} from '@envelop/core'
import { useDisableIntrospection } from '@envelop/disable-introspection'
import { useValidationCache } from '@envelop/validation-cache'
import { useParserCache } from '@envelop/parser-cache'
import { makeExecutableSchema } from '@graphql-tools/schema'
import { IResolvers, TypeSource } from '@graphql-tools/utils'

export type ServerCORSOptions = {
  origin: string[]
  methods?: string[]
  allowedHeaders?: string[]
  exposedHeaders?: string[]
  credentials?: boolean
  maxAge?: number
  optionsSuccessStatus?: number
}

const DEFAULT_CORS_OPTIONS: ServerCORSOptions = {
  origin: ['*'],
  methods: ['GET', 'HEAD', 'PUT', 'PATCH', 'POST', 'DELETE'],
  optionsSuccessStatus: 204,
}

export type YogaLogger = Pick<Console, 'debug' | 'error' | 'warn' | 'info'>

/**
 * Configuration options for the server
 */
export type ServerOptions<TContext> = {
  /**
   * Envelop Plugins
   * @see https://envelop.dev/plugins
   */
  plugins?: Array<Plugin<TContext>>
  /**
   * Enable logging
   * @default true
   */
  enableLogging?: boolean
  /**
   * Custom logger
   *
   * @default console
   */
  logger?: YogaLogger
  /**
   * Allow introspection query. This is useful for exploring the API with tools like GraphiQL.
   * If you are making a private GraphQL API,
   * it is suggested that you disable this in production so that
   * potential malicious API consumers do not see what all operations are possible.
   *
   * You can learn more about GraphQL introspection here:
   * @see https://graphql.org/learn/introspection/
   *
   * Default: `true`
   */
  introspection?: boolean
  /**
   * Prevent leaking unexpected errors to the client. We highly recommend enabling this in production.
   * If you throw `GraphQLServerError`/`EnvelopError` within your GraphQL resolvers then that error will be sent back to the client.
   *
   * You can lean more about this here:
   * @see https://www.envelop.dev/plugins/use-masked-errors
   *
   * Default: `false`
   */
  maskedErrors?: boolean | UseMaskedErrorsOpts
  /**
   * Context
   */
  context?: (req: Request) => Promise<TContext> | Promise<TContext>
  cors?: ((request: Request) => ServerCORSOptions) | ServerCORSOptions | boolean
  /**
   * GraphiQL options
   *
   * Default: `true`
   */
  graphiql?: GraphiQLOptions | boolean
} & (
  | {
      schema: GraphQLSchema
    }
  | {
      typeDefs: TypeSource
      resolvers?: IResolvers<any, TContext>
    }
)

/**
 * Base class that can be extended to create a GraphQL server with any HTTP server framework.
 * @internal
 */
export class Server<TContext> {
  /**
   * Request handler for helix
   */
  public readonly handleRequest = handleRequest
  public readonly schema: GraphQLSchema
  /**
   * Instance of envelop
   */
  public readonly getEnveloped: GetEnvelopedFn<TContext>
  public logger: YogaLogger
  public readonly corsOptionsFactory?: (request: Request) => ServerCORSOptions
  public readonly graphiql: GraphiQLOptions | false

  constructor(options: ServerOptions<TContext>) {
    this.schema =
      'schema' in options
        ? options.schema
        : makeExecutableSchema({
            typeDefs: options.typeDefs,
            resolvers: options.resolvers,
          })

    this.logger = options.enableLogging
      ? options.logger || console
      : {
          debug: () => {},
          error: () => {},
          warn: () => {},
          info: () => {},
        }

    const maskedErrors = options.maskedErrors || false

    const introspectionEnabled = options.introspection ?? true

    this.getEnveloped = envelop({
      plugins: [
        // Use the schema provided by the user
        useSchema(this.schema),
        // Performance things
        useParserCache({
          errorCache: new Map(),
          documentCache: new Map(),
        }),
        useValidationCache({
          cache: new Map(),
        }),
        // Log events - useful for debugging purposes
        enableIf(
          !!options.enableLogging,
          useLogger({
            logFn: (eventName, events) => {
              if (eventName === 'execute-start') {
                const context = events.args.contextValue
                const query = context.request?.body?.query
                const variables = context.request?.body?.variables
                const headers = context.request?.headers
                this.logger.debug(eventName)
                this.logger.debug(query, 'query')
                // there can be no variables
                if (variables && Object.keys(variables).length > 0) {
                  this.logger.debug(variables, 'variables')
                }
                this.logger.debug(headers, 'headers')
              }
              if (eventName === 'execute-end') {
                this.logger.debug(eventName)
                this.logger.debug(events.result, 'response')
              }
            },
          }),
        ),
        enableIf(!introspectionEnabled, useDisableIntrospection()),
        enableIf(
          !!maskedErrors,
          useMaskedErrors(
            typeof maskedErrors === 'object' ? maskedErrors : undefined,
          ),
        ),
        ...(options.context != null
          ? [
              useExtendContext(
                typeof options.context === 'function'
                  ? options.context
                  : () => options.context,
              ),
            ]
          : []),
        ...(options.plugins || []),
      ],
    })

    if (options.cors != null) {
      if (typeof options.cors === 'function') {
        const userProvidedCorsOptionsFactory = options.cors
        this.corsOptionsFactory = (...args) => {
          const corsOptions = userProvidedCorsOptionsFactory(...args)
          return {
            ...DEFAULT_CORS_OPTIONS,
            ...corsOptions,
          }
        }
      } else if (typeof options.cors === 'object') {
        const corsOptions = {
          ...DEFAULT_CORS_OPTIONS,
          ...options.cors,
        }
        this.corsOptionsFactory = () => corsOptions
      } else if (typeof options.cors === 'boolean') {
        this.corsOptionsFactory = () => DEFAULT_CORS_OPTIONS
      }
    }

    if (typeof options.graphiql === 'object' || options.graphiql === false) {
      this.graphiql = options.graphiql
    } else {
      this.graphiql = introspectionEnabled ? {} : false
    }
  }
}

export function createServer<TContext>(options: ServerOptions<TContext>) {
  return new Server<TContext>(options)
}

export { renderGraphiQL } from '@graphql-yoga/handler'
