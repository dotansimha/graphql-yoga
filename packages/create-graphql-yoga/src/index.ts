import ora from 'ora'
import { parseArgs } from 'node:util'
import tar from 'tar'
import { join } from 'node:path'
import { Readable } from 'node:stream'
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs'
import { createInterface } from 'node:readline'
import getNpmTarballUrl from 'get-npm-tarball-url'

export const spinner = ora()

const options = {
  template: {
    type: 'string',
    short: 't',
  },
} as const

function getRegistryAPIUrl(packageName: string, version: string) {
  return `https://registry.npmjs.org/${packageName}/${version}`
}

async function getVersionByTag(
  packageName: string,
  tag: string,
  fetchFn: typeof fetch,
) {
  const url = getRegistryAPIUrl(packageName, tag)
  const response = await fetchFn(url)
  if (response.status === 404) {
    throw new Error(`Package not found: ${packageName}`)
  }
  if (!response.ok) {
    throw new Error(
      `Failed to fetch package ${packageName} with ${
        response.status
      }: ${await response.text()}`,
    )
  }
  const { version } = await response.json()
  return version
}

function getPackageNameAndTagForTemplate(template: string) {
  const [suffix, tag] = template.split('@')
  return {
    packageName: `@graphql-yoga/template-${suffix}`,
    tag: tag ?? 'latest',
  }
}

export interface CreateGraphQLYogaOpts {
  argv: string[]
  input: Readable
  output: NodeJS.WritableStream
  fetchFn: typeof fetch
}

export async function createGraphQLYoga({
  argv,
  input,
  output,
  fetchFn,
}: CreateGraphQLYogaOpts) {
  const args = [...argv]
  while (args[0].startsWith('/') || args[0] === '--') {
    args.shift()
  }
  const rl = createInterface({
    input,
    output,
  })

  const projectName = await new Promise<string>((resolve) => {
    rl.question('What is the name of your project? ', (answer) => {
      resolve(answer)
    })
  })

  const {
    values: { template = 'node-ts' },
  } = parseArgs({ args, options, allowPositionals: true })
  spinner.start(`Fetching template ${template}...`)
  const { packageName, tag } = getPackageNameAndTagForTemplate(template)
  const version = await getVersionByTag(packageName, tag, fetchFn)
  const url = getNpmTarballUrl(packageName, version)
  const response = await fetchFn(url)
  if (response.status === 404) {
    throw new Error(`Template not found: ${template}`)
  }
  if (!response.ok) {
    throw new Error(
      `Failed to fetch template ${template} with ${
        response.status
      }: ${await response.text()}`,
    )
  }
  if (!response.body) {
    throw new Error(`Failed to fetch template ${template} with empty body`)
  }
  const nodeStream = Readable.from(
    response.body as unknown as AsyncIterable<Uint8Array>,
  )
  const targetDir = join(process.cwd(), projectName)
  if (existsSync(targetDir)) {
    throw new Error(`Target directory ${targetDir} already exists.`)
  }
  mkdirSync(targetDir, { recursive: true })
  await new Promise<void>((resolve, reject) => {
    const extractedTarStream = tar.extract({
      strip: 1,
      cwd: targetDir,
    })
    nodeStream
      .pipe(extractedTarStream)
      .once('error', (err) => {
        reject(new Error(`Failed to extract template ${template} with ${err}`))
      })
      .once('close', () => {
        resolve()
      })
  })
  const packageJsonPath = join(targetDir, 'package.json')
  const packageJsonContent = readFileSync(packageJsonPath, 'utf-8')
  const packageJson = JSON.parse(packageJsonContent)
  packageJson.name = projectName
  delete packageJson.version
  packageJson.private = true
  const newPackageJsonContent = JSON.stringify(packageJson, null, 2)
  writeFileSync(packageJsonPath, newPackageJsonContent)

  spinner.succeed(`Project ${projectName} created on ${targetDir}.`)
}
