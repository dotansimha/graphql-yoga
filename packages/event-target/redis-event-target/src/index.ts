import {
  type TypedEventTarget,
  type EventAPI,
  resolveGlobalConfig,
} from '@graphql-yoga/typed-event-target'
import type { Redis, Cluster } from 'ioredis'

export type CreateRedisEventTargetArgs = {
  publishClient: Redis | Cluster
  subscribeClient: Redis | Cluster
  /**
   * Event and EventTarget implementation.
   * Providing this is mandatory for a Node.js versions below 16.
   */
  event?: EventAPI
}

export function createRedisEventTarget<TEvent extends Event>(
  args: CreateRedisEventTargetArgs,
): TypedEventTarget<TEvent> {
  const { publishClient, subscribeClient } = args
  const eventAPI = resolveGlobalConfig(args.event)

  const callbacksForTopic = new Map<string, Set<(event: TEvent) => void>>()

  function onMessage(channel: string, message: string) {
    const callbacks = callbacksForTopic.get(channel)
    if (callbacks === undefined) {
      return
    }
    const event = new eventAPI.Event(channel) as TEvent & {
      data: unknown
    }
    event.data = message === '' ? undefined : JSON.parse(message)
    for (const callback of callbacks) {
      callback(event)
    }
  }

  subscribeClient.on('message', onMessage)

  function addCallback(topic: string, callback: (event: TEvent) => void) {
    let callbacks = callbacksForTopic.get(topic)
    if (callbacks === undefined) {
      callbacks = new Set()
      callbacksForTopic.set(topic, callbacks)

      subscribeClient.subscribe(topic)
    }
    callbacks.add(callback)
  }

  function removeCallback(topic: string, callback: (event: TEvent) => void) {
    let callbacks = callbacksForTopic.get(topic)
    if (callbacks === undefined) {
      return
    }
    callbacks.delete(callback)
    if (callbacks.size > 0) {
      return
    }
    callbacksForTopic.delete(topic)
    subscribeClient.unsubscribe(topic)
  }

  return {
    addEventListener(topic, callbackOrOptions) {
      const callback =
        'handleEvent' in callbackOrOptions
          ? callbackOrOptions.handleEvent
          : callbackOrOptions
      addCallback(topic, callback)
    },
    dispatchEvent(event: TEvent) {
      publishClient.publish(
        event.type,
        (event as any).data === undefined
          ? ''
          : JSON.stringify((event as any).data),
      )
      return true
    },
    removeEventListener(topic, callbackOrOptions) {
      const callback =
        'handleEvent' in callbackOrOptions
          ? callbackOrOptions.handleEvent
          : callbackOrOptions
      removeCallback(topic, callback)
    },
  }
}
