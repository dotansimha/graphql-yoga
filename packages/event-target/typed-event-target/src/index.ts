export type TypedEvent = Event & {
  data?: unknown
}

export interface TypedEventListener<TEvent extends TypedEvent>
  extends EventListener {
  (evt: TEvent): void
}

export interface TypedEventListenerObject<TEvent extends TypedEvent>
  extends EventListener {
  handleEvent(object: TEvent): void
}

export type TypedEventListenerOrEventListenerObject<TEvent extends TypedEvent> =
  TypedEventListener<TEvent> | TypedEventListenerObject<TEvent>

export interface TypedEventTarget<TEvent extends TypedEvent>
  extends EventTarget {
  addEventListener(
    type: string,
    callback: TypedEventListenerOrEventListenerObject<TEvent>,
    options?: AddEventListenerOptions | boolean,
  ): void
  dispatchEvent(event: TEvent): boolean
  removeEventListener(
    type: string,
    callback: TypedEventListenerOrEventListenerObject<TEvent>,
    options?: EventListenerOptions | boolean,
  ): void
}

export type EventAPI = {
  Event: typeof Event
  EventTarget: typeof EventTarget
}
