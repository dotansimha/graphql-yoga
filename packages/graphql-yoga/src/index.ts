export { createGraphQLError } from './error.js'
export * from '@graphql-yoga/logger'
export { type Plugin } from './plugins/types.js'
export { type GraphiQLOptions } from './plugins/use-graphiql.js'
export { renderGraphiQL, shouldRenderGraphiQL } from './plugins/use-graphiql.js'
export { useReadinessCheck } from './plugins/use-readiness-check.js'
export { useSchema } from './plugins/use-schema.js'
export * from './schema.js'
export * from './server.js'
export * from './subscription.js'
export * from './types.js'
export { maskError } from './utils/mask-error.js'
export { createLRUCache } from './utils/create-lru-cache.js'
export {
  // Handy type utils
  type Maybe,
  type Optional,
  type PromiseOrValue,
  type Spread,
  // useful for anyone creating a new envelop instance
  envelop,
  errorAsyncIterator,
  finalAsyncIterator,
  handleStreamOrSingleExecutionResult,
  isAsyncIterable,
  // useful helpers
  isIntrospectionOperationString,
  makeExecute,
  makeSubscribe,
  mapAsyncIterator,
  // Default plugins
  useEnvelop,
  useErrorHandler,
  useExtendContext,
  useLogger,
  usePayloadFormatter,
} from '@envelop/core'
export { getSSEProcessor } from './plugins/result-processor/sse.js'
