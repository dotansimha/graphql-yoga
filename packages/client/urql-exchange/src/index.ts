import {
  buildHTTPExecutor,
  HTTPExecutorOptions,
} from '@graphql-tools/executor-http'
import { executorExchange } from '@graphql-tools/executor-urql-exchange'
import { Exchange } from '@urql/core'

export type YogaExchangeOptions = HTTPExecutorOptions

export function yogaExchange(options?: HTTPExecutorOptions): Exchange {
  return executorExchange(buildHTTPExecutor(options as any))
}
