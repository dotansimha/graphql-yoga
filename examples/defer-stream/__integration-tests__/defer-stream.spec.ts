import { yoga } from '../src/yoga';

describe('Defer / Stream', () => {
  it('stream', async () => {
    const start = Date.now();
    const response = await yoga.fetch('/graphql', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Accept: 'application/graphql-response+json, application/json, multipart/mixed',
      },
      body: JSON.stringify({
        query: /* GraphQL */ `
          query StreamAlphabet {
            alphabet(waitFor: 100) @stream
          }
        `,
      }),
    });
    expect(response.status).toEqual(200);
    const contentType = response.headers.get('Content-Type');
    expect(contentType).toEqual('multipart/mixed; boundary="-"');
    const responseText = await response.text();
    const end = Date.now();
    expect(responseText).toMatchSnapshot('stream');
    const diff = end - start;
    expect(diff).toBeLessThan(2650);
    expect(diff > 2550).toBeTruthy();
  });
  it('defer', async () => {
    const start = Date.now();
    const response = await yoga.fetch('/graphql', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Accept: 'application/graphql-response+json, application/json, multipart/mixed',
      },
      body: JSON.stringify({
        query: /* GraphQL */ `
          query SlowAndFastFieldWithDefer {
            ... on Query @defer {
              slowField(waitFor: 1500)
            }
            fastField
          }
        `,
      }),
    });
    expect(response.status).toEqual(200);
    const contentType = response.headers.get('Content-Type');
    expect(contentType).toEqual('multipart/mixed; boundary="-"');
    const responseText = await response.text();
    const end = Date.now();
    expect(responseText).toMatchSnapshot('defer');
    const diff = end - start;
    expect(diff).toBeLessThan(1600);
    expect(diff > 1450).toBeTruthy();
  });
});
