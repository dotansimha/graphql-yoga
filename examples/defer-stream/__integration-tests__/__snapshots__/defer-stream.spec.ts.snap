// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Defer / Stream defer: defer 1`] = `
"---
Content-Type: application/json; charset=utf-8
Content-Length: 50

{"data":{"fastField":"I am speed"},"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 78

{"incremental":[{"data":{"slowField":"I am slow"},"path":[]}],"hasNext":false}
---
-----
"
`;

exports[`Defer / Stream stream: stream 1`] = `
"---
Content-Type: application/json; charset=utf-8
Content-Length: 39

{"data":{"alphabet":[]},"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 70

{"incremental":[{"items":["a"],"path":["alphabet",0]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 70

{"incremental":[{"items":["b"],"path":["alphabet",1]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 70

{"incremental":[{"items":["c"],"path":["alphabet",2]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 70

{"incremental":[{"items":["d"],"path":["alphabet",3]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 70

{"incremental":[{"items":["e"],"path":["alphabet",4]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 70

{"incremental":[{"items":["f"],"path":["alphabet",5]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 70

{"incremental":[{"items":["g"],"path":["alphabet",6]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 70

{"incremental":[{"items":["h"],"path":["alphabet",7]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 70

{"incremental":[{"items":["i"],"path":["alphabet",8]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 70

{"incremental":[{"items":["j"],"path":["alphabet",9]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 71

{"incremental":[{"items":["k"],"path":["alphabet",10]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 71

{"incremental":[{"items":["l"],"path":["alphabet",11]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 71

{"incremental":[{"items":["m"],"path":["alphabet",12]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 71

{"incremental":[{"items":["n"],"path":["alphabet",13]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 71

{"incremental":[{"items":["o"],"path":["alphabet",14]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 71

{"incremental":[{"items":["p"],"path":["alphabet",15]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 71

{"incremental":[{"items":["q"],"path":["alphabet",16]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 71

{"incremental":[{"items":["r"],"path":["alphabet",17]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 71

{"incremental":[{"items":["s"],"path":["alphabet",18]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 71

{"incremental":[{"items":["t"],"path":["alphabet",19]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 71

{"incremental":[{"items":["u"],"path":["alphabet",20]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 71

{"incremental":[{"items":["v"],"path":["alphabet",21]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 71

{"incremental":[{"items":["w"],"path":["alphabet",22]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 71

{"incremental":[{"items":["x"],"path":["alphabet",23]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 71

{"incremental":[{"items":["y"],"path":["alphabet",24]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 71

{"incremental":[{"items":["z"],"path":["alphabet",25]}],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 17

{"hasNext":false}
---
-----
"
`;
